### ISSUE TITLE
feature request: Make Epic run in webworker if wanted

### ISSUE THREAD

**Issue by nasreddineskandrani** [redux-observable/redux-observable#278]
Hi,
I want the input of the guys behind redux-observable about the ability to have an epic that run in a 'web-worker' built in in the redux-observable lib?
If you think its faisable how would you make it? we can help to implement this if needed.

Thank you

**Comment by jayphelps** [comment]
Yeah it would totally be doable--easiest way would be your root epic on the main thread maintains the webworker and proxies actions to and from the worker via postMessage. Since actions are _supposed_ to be serializable that part shouldn't be an issue. The API for `store.getState()` would have to be customized to be async, though.

I see you just edited your post to say "build in the lib"..I think you mean having this ability built-in to redux-observable. If that's the case, it's not likely since a large majority of people wouldn't need/use it but it would be easy enough as an optional addon.

Why do you want to run them in a webworker? It's unlikely to provide any performance benefit (likely opposite because of serialization) except in very particular use cases which would warrant one-off stuff anyway. At least, in my experience. 😄  Maybe you have something in mind?

**Comment by nasreddineskandrani** [comment]
Most of the time, an epic in my case is doing async job as grabbing data from backend. This data can be huge sometimes as geojson etc. Dealing with this in IE11 to load this in memory will freeze the js UI thread.
 Since your library is dealing with async jobs in Epics. I thought why not having the possibility to have an epic on a WebWorker to don't code that outside.
I was thinking about indexes db to read/write the data received and don't use postMessage.
I ll try to build an example with some problematic use cases that webworker can solve. 
But for sure its not a mandatory piece in redux-observable lib.

**Comment by nasreddineskandrani** [comment]
sorry i didn't forget... was in a middle of releasing my first product + my day to day job. Maybe next weekend i ll really try to show a usecase. I wonder if Edge has the problem i am mentionning with IE.

**Comment by huy-nguyen** [comment]
I actually built a web app (http://atlas.cid.harvard.edu) that runs most epics for data fetching inside a web worker.

**Comment by nasreddineskandrani** [comment]
on top of my job, i invested my time in lazy-loading react/redux-observable check last post here: https://github.com/redux-observable/redux-observable/issues/365
and i am also working in the same topic in angular side comparing with React.
https://github.com/ngrx/platform/issues/972

so i didn't have time to check this webworker topic yet. Amazin that you used webworkers. 
questions:
- What lead you to use web-worker? any technical blocker?
- Do you have any open source example?

**Comment by AaronVasquez** [comment]
@huy-nguyen any way you can share how you did that? Would love to have this feature with epics.

**Comment by huy-nguyen** [comment]
@AaronVasquez I'll try to make a minimal example repo with an accompanying blog post and post it here.

**Comment by slothyrulez** [comment]
@huy-nguyen that would be awesome

**Comment by nwsrh** [comment]
@huy-nguyen Could you give an example of how you did it, please. I'm developing a project that requires a lot of processing and this idea I think would help me a lot. I would really appreciate it.

**Comment by huy-nguyen** [comment]
Oops sorry I didn't keep my promise made last year. I'll try to find time to that in the next week or two.
